build.wake

```diff
@@ -32,6 +32,27 @@ def firrtlProto Unit =
   | runJob
   | getJobOutputs
 
+def firrtlVersion = "1.4.0-RC3"
+
+def buildInfo ver =
+  def content =
+    "%
+      package firrtl
+      import scala.Predef._
+      /** This object was generated by api-firrtl-sifive */
+      case object BuildInfo {
+        val buildInfoPackage: String = "firrtl"
+        val version: String = "%{firrtlVersion}"
+        val scalaVersion: String = "%{ver.scalaVersionToString}"
+        override val toString: String =
+          s"buildInfoPackage: $buildInfoPackage, version: $version, scalaVersion: $scalaVersion"
+      }
+      %"
```

Comment:
This approach is an adaptation of https://github.com/sbt/sbt-buildinfo which is how we've done it in Chisel forever (and FIRRTL's SBT build is using that plugin as well). It's DRYer than having to maintain a separate source file.

Reply:
Maybe I'm not seeing the value of DRYness given that this already duplicates information that is already described in firrtl.git's build.sbt.

But even if I accept that we need to dynamically generate this, the way this is written still bothers me, since it's in an awkward middle ground between "general" and "hardcoding specific things". From my perspective as a reader of this code, the first thing I see is the function declaration, which makes it look like a `buildInfo` is a function that takes in a `ver` argument. That signals to me that "this is something that can be called with different arguments".

However, reading through the rest of the function's definition, I can see a hardcoded value for the output file path (it closes over the `buildDir` variable, which is a constant defined elsewhere in this file). If you try calling this function with two different arguments, it'll hit a Wake error due to two invocations writing to the same file path. This means that this function cannot in fact called twice, at least within the same Wake process.

This is surprising to me, as a reader of the code, since I can no longer compartmentalize this into something that is "general" or something that is "specific", since the level of flexibility of the function argument is inconsistent with the level of flexibility of its output path. Because the function takes an argument, it implies that you can call it more than once, but that's not actually true, and you'll get a runtime error if you do so.

To me, it would make more sense for this function to either 1) not accept any parameters (besides Unit), instead grabbing the Scala version from another file-scope singleton object for the Scala module, or 2) allow the firrtl version and the output path be parameters to the function.

Sorry, this entire post is probably a huge bikeshed, since it's ultimately inconsequential, but I wanted to share why I'm getting weird vibes from it.

Reply:
> Maybe I'm not seeing the value of DRYness given that this already duplicates information that is already described in firrtl.git's build.sbt.

That's a case where it's incredibly impractical to be DRY (SBT can't read that kind of information from a file), whereas this is a case where we can avoid it.

I agree with your criticism though that the function is neither general nor specific. I think it does make sense to write this in a more generic way, then it can eventually be spun out elsewhere and in the meantime can be used by `api-chsiel3-sifive` (since this logic is currently duplicated in each project 😉 )